### PR TITLE
Add v1/:network/tokens/:id endpoint

### PR DESCRIPTION
**Pull request by jingyi2811**
/kind feature

Add v1/:network/tokens related rpcs

### CODE REVIEW

File: src/module.api/tokens.controller.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  /**
+   * @param {string} id token's symbol key
+   * @return {TokenResult}
+   */
+  @Get('/:id')
+  async get (@Param('id') id: string): Promise<TokenResult> {
+    try {
+      return await this.client.token.getToken(id)
+    } catch (e) {
+      throw new BadRequestException()
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.ts:23]: You need to test this aspect in E2E, all edge/validation cases must be tested.

File: src/module.api/tokens.controller.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  isDAT: boolean
+  isLPS: boolean
+  finalized: boolean
+  minted: number
+  creationTx: string
+  creationHeight: number
+  destructionTx: string
+  destructionHeight: number
+  collateralAddress: string
+}
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.ts:49]: You need to implement this: https://github.com/DeFiCh/whale/pull/24#discussion_r624917549

Comment (fuxingloh) [reply src/module.api/tokens.controller.ts:49]: Whale service should never be a carbon copy of what the RPC throws out.

File: src/module.api/tokens.controller.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+export class TokensController {
+  constructor (private readonly client: JsonRpcClient) {
+  }
+
+  /**
+   * @param {string} id token's symbol key
+   * @return {TokenResult}
+   */
+  @Get('/:id')
+  async get (@Param('id') id: string): Promise<TokenResult> {
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.ts:19]: This is a singular REST endpoint/contract, you put out one request for a `/:id` and expect one piece of data back. Not a map of thme.

Hence you should not return `Promise<TokenResult>`.
You need to return a singular `TokenInfo`, if the RPC does not conform to REST, you service need to.

File: __e2e__/tokens.e2e.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import { NestFastifyApplication } from '@nestjs/platform-fastify'
+
+const container = new MasterNodeRegTestContainer()
+let app: NestFastifyApplication
+
+beforeAll(async () => {
+  await container.start()
+  await container.waitForReady()
+  app = await createTestingApp(container)
+})
```
Comment (fuxingloh) [inline __e2e__/tokens.e2e.ts:12]: You need to setup at least another condition to test another token.

File: __e2e__/tokens.e2e.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+afterAll(async () => {
+  await container.stop()
+})
+
+describe('GET: /v1/regtest/tokens/DFI', () => {
```
Comment (fuxingloh) [inline __e2e__/tokens.e2e.ts:18]: 1. Test all expected errors, validation, malformed, bad request
2. Test other endpoints not just one /:id

**Review by fuxingloh — CHANGES_REQUESTED** [review]

File: src/module.api/tokens.controller.spec.ts [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+  is_dat: boolean
+  is_lps: boolean
+  finalized: boolean
+  minted: number
+  creation_tx: string
+  creation_height: number
+  destruction_tx: string
+  destruction_height: number
+  collateral_address: string
+}
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.spec.ts:52]: should not need to export new interfaces for tests, can test blindly with `any` since your test will guarantee the values exists. as you will be asserting them. (remember typescript is compile-time type-safety only if you throw a JSON parsed value, it won't know the features of that parsed value)

File: src/module.api/tokens.controller.ts (hunk 2 above)
Comment (fuxingloh) [reply src/module.api/tokens.controller.ts:49]: If you need to use interface from jellyfish you should import it, instead of copying it: https://github.com/DeFiCh/whale/pull/24#discussion_r626376928

File: src/module.api/tokens.controller.spec.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+})
+
+describe('controller.get() for token which does not exist', () => {
+  it('should return Token not found with id as param', async () => {
+    try {
+      await controller.get('2')
+    } catch (e) {
+      expect(e.message).toBe('Token not found')
+    }
+  })
```
Comment (canonbrother) [inline src/module.api/tokens.controller.spec.ts:164]: hmm.. this is not quite right as the `ln: 162` might not 100% get triggered
jest doesn't have assert.fail (mocha) but jest can do expect(func()).throw(err)

File: src/module.api/tokens.controller.spec.ts [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+      load: [() => ({ network: 'regtest' })]
+    })],
+    controllers: [TokensController],
+    providers: [{ provide: JsonRpcClient, useValue: client }]
+  }).compile()
+
+  controller = app.get<TokensController>(TokensController)
+})
+
+export interface TokenInfoDto {
```
Comment (canonbrother) [inline src/module.api/tokens.controller.spec.ts:35]: ohh.. its duplicated.. just grab it from the controller

File: __e2e__/tokens.e2e.ts [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+    symbol,
+    name: symbol,
+    isDAT: true,
+    mintable: true,
+    tradeable: true,
+    collateralAddress: address
+  }
+  await container.call('createtoken', [metadata])
+  await container.generate(1)
+}
```
Comment (fuxingloh) [inline __e2e__/tokens.e2e.ts:31]: rebase main, use the new `@defichain/testing` for testing

`npm i --save-dev @defichain/testing@next`

**Review by fuxingloh — CHANGES_REQUESTED** [review]

File: src/module.api/tokens.controller.ts [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+  /**
+   * Returns information about token.
+   *
+   * @param {number} size Maximum number of tokens to return.
+   * @param {number} next first key to iterate from, in lexicographical order.
+   * @return {Promise<SliceResponse<TokenInfoDto>>}
+   */
+  @Get('/')
+  async get (
+    @Query(new TokenSizePipe()) size = '1000000000',
```
Comment (jingyi2811) [inline src/module.api/tokens.controller.ts:87]: This value 1000000000 looks a bit funny. But, I was trying to select every tokens

File: src/module.api/tokens.controller.ts [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+      const data = await this.client.token.listTokens(pagination, true)
+
+      const tokenInfoDtos: TokenInfoDto[] = []
+
+      for (const key of Object.keys(data)) {
+        tokenInfoDtos.push(toTokenInfoDTO(data[key]))
+      }
+
+      return SliceResponse.of(tokenInfoDtos, limit, item => {
```
Comment (jingyi2811) [inline src/module.api/tokens.controller.ts:108]: Not too sure whether my implementation is correct or not.

**Comment by jingyi2811** [comment]
Hope someone can review my code.

File: src/module.api/tokens.controller.ts (hunk 10 above)
Comment (canonbrother) [reply src/module.api/tokens.controller.ts:87]: why is that a string on size??

Comment (canonbrother) [reply src/module.api/tokens.controller.ts:87]: you don't need TokenSizePipe as that's no extra remap work
see https://docs.nestjs.com/techniques/validation

File: src/module.api/tokens.controller.ts [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+   * Returns information about token.
+   *
+   * @param {number} size Maximum number of tokens to return.
+   * @param {number} next first key to iterate from, in lexicographical order.
+   * @return {Promise<SliceResponse<TokenInfoDto>>}
+   */
+  @Get('/')
+  async get (
+    @Query(new TokenSizePipe()) size = '1000000000',
+    @Query(new TokenNextPipe()) next = '0'
```
Comment (canonbrother) [inline src/module.api/tokens.controller.ts:88]: same, don't need `TokenNextPipe` here

File: src/module.api/tokens.controller.ts [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+/**
+ * Map TokenInfo to TokenInfoDto.
+ *
+ * @param {TokenInfo} tokenInfo
+ * @return {TokenInfoDto}
+ */
+function toTokenInfoDTO (tokenInfo: TokenInfo): TokenInfoDto {
+  return {
+    symbol: tokenInfo.symbol,
+    symbol_key: tokenInfo.symbolKey,
```
Comment (canonbrother) [inline src/module.api/tokens.controller.ts:142]: camelCase

**Review by canonbrother — CHANGES_REQUESTED** [review]

File: packages/whale-api-client/src/api/tokens.ts [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  /**
+   * Returns information about tokens.
+   *
+   * @param {number} size of tokens to query
+   * @param {number} next set of tokens
+   * @return {Promise<ApiPagedResponse<TokenInfo>>}
+   */
+  async get (size: number = 30, next?: string): Promise<ApiPagedResponse<TokenInfo>> {
```
Comment (canonbrother) [inline packages/whale-api-client/src/api/tokens.ts:16]: list

File: src/module.api/tokens.controller.ts [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  /**
+   * Returns information about tokens.
+   *
+   * @param {string} id id/symbol/creationTx
+   * @return {Promise<TokenInfo>}
+   */
+  @Get('/:id')
+  async getId (@Param('id') id: string): Promise<TokenInfo> {
```
Comment (canonbrother) [inline src/module.api/tokens.controller.ts:61]: get

File: src/module.api/tokens.controller.ts [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+    const data = await this.client.token.getToken(id)
```
Comment (canonbrother) [inline src/module.api/tokens.controller.ts:62]: do cache the token here

File: src/module.api/tokens.controller.ts [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+  decimal: number
+  limit: number
+  mintable: boolean
+  tradeable: boolean
+  isDAT: boolean
+  isLPS: boolean
+  finalized: boolean
+  minted: number
+  creationTx: string
+  creationHeight: number
```
Comment (canonbrother) [inline src/module.api/tokens.controller.ts:81]: group the same category property
eg: 
```
creation: {
  tx,
  height,
}

**Review by canonbrother — CHANGES_REQUESTED** [review]
code looks cleaner a lot than before

File: src/module.api/tokens.controller.ts [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+          isLPS: value.isLPS,
+          finalized: value.finalized,
+          minted: value.minted,
+          creationTx: value.creationTx,
+          creationHeight: value.creationHeight,
+          destructionTx: value.destructionTx,
+          destructionHeight: value.destructionHeight,
+          collateralAddress: value.collateralAddress
+        }
+      }).sort(a => Number.parseInt(a.id))
```
[the remark this answers is not in the document]
Comment (fuxingloh) [reply src/module.api/tokens.controller.ts:48]: see AddressController for more info.

File: src/module.api/tokens.controller.ts [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+    return ApiPagedResponse.of(tokens, query.size, item => {
+      return item.id
+    })
+  }
+
+  /**
+   * Returns information about tokens.
+   *
+   * @param {string} id id/symbol/creationTx
```
[the remark this answers is not in the document]
Comment (fuxingloh) [reply src/module.api/tokens.controller.ts:57]: let's not support `symbol` or `creationTx`, DeFi Whale don't need to follow Ain. We can keep the RESTful resources construct.

File: src/module.api/tokens.controller.ts (hunk 16 above)
Comment (fuxingloh) [reply src/module.api/tokens.controller.ts:62]: For cache, it is optional now, but a bonus if you implement it. Don't have to yet, you can create a ticket for it.

File: src/module.api/tokens.controller.ts [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+import { TokenInfo } from '@defichain/jellyfish-api-core/dist/category/token'
+import { PaginationQuery } from '@src/module.api/_core/api.query'
+
+@Controller('/v1/:network/tokens')
+export class TokensController {
+  constructor (private readonly client: JsonRpcClient) {
+  }
+
+  /**
+   * Returns information about tokens.
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.ts:13]:
```suggestion
   * Paginate query tokens.
```

File: src/module.api/tokens.controller.spec.ts [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+    await expect(controller.getId('MOCK'))
+      .rejects
+      .toThrow('Token not found')
+  })
+
+  it('should fail with creationTx as param', async () => {
+    await expect(controller.getId('5000000000000000000000000000000000000000000000000000000000000000'))
+      .rejects
+      .toThrow('Token not found')
+  })
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.spec.ts:204]: these are ain/rpc design pattern, as mentioned, we only need to support RESTful API Design.

File: src/module.api/tokens.controller.spec.ts [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+    providers: [{ provide: JsonRpcClient, useValue: client }]
+  }).compile()
+  controller = app.get<TokensController>(TokensController)
+})
+
+describe('controller.get() for all tokens', () => {
+  it('should listTokens', async () => {
+    const result = await controller.get({ size: 100 })
+
+    result.data.forEach(function (data) {
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.spec.ts:35]: It will always be ordered as `[0]`, `[1]` use `[n]` to access array instead of forEach. ForEach will not be looped if empty hence your test is flawed when it return an empty array

File: packages/whale-api-client/src/api/tokens.ts [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+    return await this.client.requestList('GET', 'tokens', size, next)
+  }
+
+  /**
+   * Returns information about tokens.
+   *
+   * @param {string} id id/symbol/creationTx
+   * @return {Promise<TokenInfo>}
+   */
+  async getId (id: string): Promise<TokenInfo> {
```
Comment (fuxingloh) [inline packages/whale-api-client/src/api/tokens.ts:26]: get

File: packages/whale-api-client/__tests__/api/tokens.test.ts [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+  }
+})
+
+describe('client.tokens.get()', () => {
+  it('should listTokens', async () => {
+    const result = await client.tokens.get()
+    expect(result.length).toBe(2)
+    for (const k in result) {
+      const data = result[k]
+      switch (data.symbol) {
```
Comment (fuxingloh) [inline packages/whale-api-client/__tests__/api/tokens.test.ts:38]: same for this

**Review by fuxingloh — CHANGES_REQUESTED** [review]

File: src/module.api/tokens.controller.ts (hunk 16 above)
Comment (jingyi2811) [reply src/module.api/tokens.controller.ts:62]: Will do this in another ticket

Comment (jingyi2811) [reply src/module.api/tokens.controller.ts:45]: Created a ticket for this. https://github.com/DeFiCh/whale/issues/85

File: src/module.api/tokens.controller.ts [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+  /**
+   * Get information about a token with id of the token.
+   *
+   * @param {string} id
+   * @return {Promise<TokenData>}
+   */
+  @Get('/:id')
+  async get (@Param('id', ParseIntPipe) id: string): Promise<TokenData> {
+    const data = await this.client.token.getToken(id)
+    return mapTokenData('0', data[Object.keys(data)[0]])
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.ts:46]: `'0'`?

your assertions didn't catch this mistake

Comment (fuxingloh) [reply src/module.api/tokens.controller.ts:46]: How about if id is invalid?

File: src/module.api/tokens.controller.ts [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+      including_start: query.next === undefined,
+      limit: query.size
+    }, true)
+    const tokens: TokenData[] = Object.entries(data)
+      .map(([id, value]): TokenData => {
+        return mapTokenData(id, value)
+      }).sort(a => Number.parseInt(a.id))
+    return ApiPagedResponse.of(tokens, query.size, item => {
+      return item.id
+    })
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.ts:34]: add line breaks for code chunks like this

File: src/module.api/tokens.controller.ts [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+  constructor (private readonly client: JsonRpcClient) {
+  }
+
+  /**
+   * Paginate query tokens.
+   *
+   * @param {PaginationQuery} query
+   * @return {Promise<ApiPagedResponse<TokenData>>}
+   */
+  @Get('/')
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.ts:19]: Remove `'/'`

File: src/module.api/tokens.controller.spec.ts [hunk 28]
```diff
@@ -0,0 +1,161 @@
...
+
+describe('controller.list() for all tokens', () => {
+  it('should listTokens', async () => {
+    const result = await controller.list({ size: 100 })
+
+    for (let i = 0; i < result.data.length; i++) {
+      const data = result.data[i]
+      switch (data.symbol) {
+        case 'DFI':
+          expect(data.symbol).toBe('DFI')
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.spec.ts:39]: please look at the previous comment and format your test, this list is ordered.

File: packages/whale-api-client/__tests__/api/tokens.test.ts [hunk 29]
```diff
@@ -0,0 +1,146 @@
...
+    expect(data.destruction.tx).toBe('0000000000000000000000000000000000000000000000000000000000000000')
+    expect(data.destruction.height).toBe(-1)
+    expect(typeof data.collateralAddress).toBe('string')
+  })
+
+  it('should fail due to id is malformed', async () => {
+    const call = async (): Promise<TokenData> => await client.tokens.get('$*@')
+    await expect(call).rejects
+      .toThrow('400 - BadRequest (/v1/regtest/tokens/$*@)')
+  })
```
Comment (fuxingloh) [inline packages/whale-api-client/__tests__/api/tokens.test.ts:145]: Add another NotFound test

File: packages/whale-api-client/__tests__/api/tokens.test.ts [hunk 30]
```diff
@@ -0,0 +1,146 @@
...
+    expect(third.length).toBe(0)
+    expect(third.hasNext).toBe(false)
+    expect(third.nextToken).toBeUndefined()
+  })
+})
+
+describe('client.tokens.get()', () => {
+  it('should return DFI coin with id as param', async () => {
+    const data = await client.tokens.get('0')
+    expect(data.symbol).toBe('DFI')
```
Comment (fuxingloh) [inline packages/whale-api-client/__tests__/api/tokens.test.ts:103]: please test `data.id` as well.

File: packages/whale-api-client/__tests__/api/tokens.test.ts [hunk 31]
```diff
@@ -0,0 +1,146 @@
...
+
+describe('client.tokens.list()', () => {
+  it('should listTokens', async () => {
+    const result = await client.tokens.list()
+    expect(result.length).toBe(2)
+    for (let i = 0; i < result.length; i++) {
+      const data = result[i]
+      switch (data.symbol) {
+        case 'DFI':
+          expect(data.symbol).toBe('DFI')
```
Comment (fuxingloh) [inline packages/whale-api-client/__tests__/api/tokens.test.ts:40]: read the previous comment about this. expecting a ordered list hence assertions should be oreded.

File: packages/whale-api-client/__tests__/api/tokens.test.ts [hunk 32]
```diff
@@ -0,0 +1,146 @@
...
+beforeAll(async () => {
+  container = new MasterNodeRegTestContainer()
+  service = new StubService(container)
+  client = new StubWhaleApiClient(service)
+
+  await container.start()
+  await container.waitForReady()
+  await container.waitForWalletCoinbaseMaturity()
+  await service.start()
+  await createToken(container, 'DSWAP')
```
Comment (fuxingloh) [inline packages/whale-api-client/__tests__/api/tokens.test.ts:21]: 1. Add 1 more token with different conditions for testing
1. Add a LP token also.

**Review by fuxingloh — CHANGES_REQUESTED** [review]
much cleaner now, a few more things to address.

File: packages/whale-api-client/__tests__/api/tokens.test.ts [hunk 33]
```diff
@@ -0,0 +1,297 @@
...
+
+    expect(third.length).toBe(1)
+    expect(third.hasNext).toBe(true)
+    expect(third.nextToken).toBe('2')
+
+    const forth = await client.paginate(third)
+
+    expect(forth.length).toBe(1)
+    expect(forth.hasNext).toBe(true)
+    expect(forth.nextToken).toBe('3')
```
Comment (canonbrother) [inline packages/whale-api-client/__tests__/api/tokens.test.ts:167]: duplicated test
please adjust `size` as long as it covers  first -> next -> last

File: packages/whale-api-client/__tests__/api/tokens.test.ts [hunk 34]
```diff
@@ -0,0 +1,297 @@
...
+        height: -1
+      },
+      collateralAddress: expect.any(String)
+    })
+  })
+
+  it('should fail due to id is not found', async () => {
+    const call = async (): Promise<TokenData> => await client.tokens.get('4')
+    await expect(call).rejects
+      .toThrow('400 - BadRequest (/v1/regtest/tokens/4)')
```
Comment (fuxingloh) [inline packages/whale-api-client/__tests__/api/tokens.test.ts:289]: This should be not found if it's not found, not bad request.

File: packages/whale-api-client/__tests__/api/tokens.test.ts [hunk 35]
```diff
@@ -0,0 +1,297 @@
...
+      destruction: {
+        tx: '0000000000000000000000000000000000000000000000000000000000000000',
+        height: -1
+      },
+      collateralAddress: ''
+    })
+
+    data = result[1]
+
+    expect(data).toEqual({
```
Comment (canonbrother) [inline packages/whale-api-client/__tests__/api/tokens.test.ts:67]: test one will do unless that has diff condition to test

**Review by canonbrother — CHANGES_REQUESTED** [review]
remove duplicate test

File: src/module.api/tokens.controller.ts [hunk 36]
```diff
@@ -0,0 +1,70 @@
...
+
+  /**
+   * Get information about a token with id of the token.
+   *
+   * @param {string} id
+   * @return {Promise<TokenData>}
+   */
+  @Get('/:id')
+  async get (@Param('id', ParseIntPipe) id: string): Promise<TokenData> {
+    const data = await this.client.token.getToken(id)
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.ts:47]: You need to catch the error to handle as NotFound.

File: src/module.api/tokens.controller.spec.ts [hunk 37]
```diff
@@ -0,0 +1,306 @@
...
+      collateralAddress: expect.any(String)
+    })
+  })
+})
+
+describe('controller.get() for token which is not found', () => {
+  it('should fail with id as param', async () => {
+    await expect(controller.get('4'))
+      .rejects
+      .toThrow('Token not found')
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.spec.ts:304]: and test for NotFoundException

Comment (fuxingloh) [reply src/module.api/tokens.controller.spec.ts:304]: see @canonbrother PR

File: src/module.api/tokens.controller.spec.ts [hunk 38]
```diff
@@ -0,0 +1,306 @@
...
+      destruction: {
+        tx: '0000000000000000000000000000000000000000000000000000000000000000',
+        height: -1
+      },
+      collateralAddress: expect.any(String)
+    })
+  })
+
+  it('should listTokens with pagination', async () => {
+    const first = await controller.list({ size: 1 })
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.spec.ts:144]: Same as e2e comment

File: src/module.api/tokens.controller.spec.ts [hunk 39]
```diff
@@ -0,0 +1,306 @@
...
+  }).compile()
+  controller = app.get<TokensController>(TokensController)
+})
+
+describe('controller.list() for all tokens', () => {
+  it('should listTokens', async () => {
+    const result = await controller.list({ size: 100 })
+    expect(result.data.length).toBe(4)
+
+    let data = result.data[0]
```
Comment (fuxingloh) [inline src/module.api/tokens.controller.spec.ts:38]: same as e2e comment.

**Review by fuxingloh — CHANGES_REQUESTED** [review]

File: packages/whale-api-client/__tests__/api/tokens.test.ts (hunk 34 above)
Comment (jingyi2811) [reply packages/whale-api-client/__tests__/api/tokens.test.ts]: Screenshot 2021-05-20 at 5 04 14 PM

Webster was able to catch WhaleApiException but I could not. Kindly advise. Thanks.

File: src/module.api/tokens.controller.ts [hunk 40]
```diff
@@ -0,0 +1,74 @@
...
+   * Get information about a token with id of the token.
+   *
+   * @param {string} id
+   * @return {Promise<TokenData>}
+   */
+  @Get('/:id')
+  async get (@Param('id', ParseIntPipe) id: string): Promise<TokenData> {
+    try {
+      const data = await this.client.token.getToken(id)
+      return mapTokenData(String(id), data[Object.keys(data)[0]])
```
Comment (jingyi2811) [inline src/module.api/tokens.controller.ts:49]: Casting to String because ParseIntPipe change the id to integer

File: packages/whale-api-client/__tests__/api/tokens.test.ts [hunk 41]
```diff
@@ -0,0 +1,124 @@
...
+
+    expect(next.length).toBe(2)
+    expect(next.hasNext).toBe(true)
+    expect(next.nextToken).toBe('3')
+
+    const last = await client.paginate(next)
+
+    expect(last.length).toBe(0)
+    expect(last.hasNext).toBe(false)
+    expect(last.nextToken).toBeUndefined()
```
Comment (fuxingloh) [inline packages/whale-api-client/__tests__/api/tokens.test.ts:81]: you need to test that data is return in an order see in `test.ts`, you did it for `spec.ts` but not in `test.ts`

**Review by fuxingloh — CHANGES_REQUESTED** [review]
1 minor change

File: src/module.api/tokens.controller.spec.ts [hunk 42]
```diff
@@ -0,0 +1,288 @@
...
+        height: 0
+      },
+      destruction: {
+        tx: '0000000000000000000000000000000000000000000000000000000000000000',
+        height: -1
+      },
+      collateralAddress: ''
+    })
+
+    expect(result.data[1]).toEqual({
```
Comment (canonbrother) [inline src/module.api/tokens.controller.spec.ts:63]: remove the duplicated here

File: src/module.api/tokens.controller.spec.ts [hunk 43]
```diff
@@ -0,0 +1,288 @@
...
+    expect(first.data.length).toBe(2)
+    expect(first.page?.next).toBe('1')
+
+    expect(first.data[0]).toEqual(expect.objectContaining({ id: '0', symbol: 'DFI', symbolKey: 'DFI' }))
+    expect(first.data[1]).toEqual(expect.objectContaining({ id: '1', symbol: 'DBTC', symbolKey: 'DBTC' }))
+
+    const next = await controller.list({
+      size: 2,
+      next: first.page?.next
+    })
```
Comment (canonbrother) [inline src/module.api/tokens.controller.spec.ts:148]: should test the `controller.paginate` here

File: src/module.api/tokens.controller.spec.ts [hunk 44]
```diff
@@ -0,0 +1,288 @@
...
+      },
+      destruction: {
+        tx: '0000000000000000000000000000000000000000000000000000000000000000',
+        height: -1
+      },
+      collateralAddress: ''
+    })
+  })
+
+  it('should return DBTC token with id as param', async () => {
```
Comment (canonbrother) [inline src/module.api/tokens.controller.spec.ts:201]: duplicated

File: src/module.api/tokens.controller.spec.ts [hunk 45]
```diff
@@ -0,0 +1,288 @@
...
+      },
+      destruction: {
+        tx: '0000000000000000000000000000000000000000000000000000000000000000',
+        height: -1
+      },
+      collateralAddress: expect.any(String)
+    })
+  })
+
+  it('should return DETH token with id as param', async () => {
```
Comment (canonbrother) [inline src/module.api/tokens.controller.spec.ts:228]: duplicated

File: src/module.api/tokens.controller.spec.ts [hunk 46]
```diff
@@ -0,0 +1,288 @@
...
+      },
+      destruction: {
+        tx: '0000000000000000000000000000000000000000000000000000000000000000',
+        height: -1
+      },
+      collateralAddress: expect.any(String)
+    })
+  })
+
+  it('should return DBTC-DETH LP token with id as param', async () => {
```
Comment (canonbrother) [inline src/module.api/tokens.controller.spec.ts:255]: duplicated

File: src/module.api/tokens.controller.ts (hunk 40 above)
Comment (canonbrother) [reply src/module.api/tokens.controller.ts:49]: `ParseIntPipe` won't convert string to int

**Review by canonbrother — CHANGES_REQUESTED** [review]

File: packages/whale-api-client/__tests__/api/tokens.test.ts (hunk 41 above)
Comment (jingyi2811) [reply packages/whale-api-client/__tests__/api/tokens.test.ts:81]: Ya, I purposely removed it. Thinking that I had already verified in spec. Let me add it back.

File: src/module.api/tokens.controller.spec.ts (hunk 43 above)
Comment (jingyi2811) [reply src/module.api/tokens.controller.spec.ts:148]: But the problem is, all other controller.spec files do not have the paginate keywords

File: packages/whale-api-client/__tests__/api/tokens.test.ts (hunk 41 above)
Comment (fuxingloh) [reply packages/whale-api-client/__tests__/api/tokens.test.ts:81]: this is because order might change when serialized, this just e2e test this behavior that a correct data structure is used.

File: src/module.api/tokens.controller.spec.ts (hunk 46 above)
Comment (fuxingloh) [reply src/module.api/tokens.controller.spec.ts:255]: We can keep LP token

File: src/module.api/tokens.controller.spec.ts (hunk 43 above)
Comment (fuxingloh) [reply src/module.api/tokens.controller.spec.ts:148]: likely type from @canonbrother since this is spec.ts

**Comment by jingyi2811** [comment]
ParseIntPipe won't convert string to int

Nvm, I follow your implementation in poolpair. You expect an integer here.

Screenshot 2021-05-20 at 7 46 10 PM

File: src/module.api/tokens.controller.spec.ts (hunk 45 above)
Comment (fuxingloh) [reply src/module.api/tokens.controller.spec.ts:228]: we can keep this since it's added

File: src/module.api/tokens.controller.spec.ts (hunk 44 above)
Comment (fuxingloh) [reply src/module.api/tokens.controller.spec.ts:201]: we can keep this since it's added

**Comment by fuxingloh** [comment]
> ParseIntPipe won't convert string to int
> 
> Nvm, I follow your implementation in poolpair. You expect an integer here.
> 
> Screenshot 2021-05-20 at 7 46 10 PM

If this is the case, write a test to assert what it should be.

File: src/module.api/tokens.controller.spec.ts (hunk 43 above)
Comment (canonbrother) [reply src/module.api/tokens.controller.spec.ts:148]: yea my bad.. seeing it wrongly

File: src/module.api/tokens.controller.ts (hunk 40 above)
Comment (canonbrother) [reply src/module.api/tokens.controller.ts:49]: tested on the `cast` and i was misunderstand `ParseIntPipep` previously as i did check the id type inside scope

File: src/module.api/tokens.controller.spec.ts (hunk 42 above)
Comment (jingyi2811) [reply src/module.api/tokens.controller.spec.ts:63]: I will keep this too

File: packages/whale-api-client/__tests__/api/poolpair.test.ts [hunk 47]
```diff
@@ -135,7 +135,7 @@ describe('get', () => {
     const response = await client.poolpair.get('7')
 
     expect(response).toEqual({
```
Comment (jingyi2811) [inline packages/whale-api-client/__tests__/api/poolpair.test.ts:137]: I changed to toStrictEqual just now, however it raised testing error, and hence,  I reverted back the code. I will do this tomorrow again.

**Review by fuxingloh — APPROVED** [review]

**Review by canonbrother — APPROVED** [review]
